src/java.base/share/classes/sun/security/provider/SecureRandom.java

```diff
@@ -190,6 +204,46 @@ private static void updateState(byte[] state, byte[] output) {
         }
     }
 
+    @Override
+    public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+        objLock.lock();
+        if (clearStateOnCheckpoint) {
+            if (state != null) {
+                for (int i = 0; i < state.length; i++) {
+                    state[i] = 0;
+                }
+            }
+            state = null;
+            if (remainder != null) {
+                for (int i = 0; i < remainder.length; i++) {
+                    remainder[i] = 0;
+                }
+            }
+            remainder = null;
+            remCount = 0;
+        }
+    }
+
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws Exception {
+        try {
+            if (isSeedGenerator) {
+                byte[] b = new byte[DIGEST_SIZE];
+                SeedGenerator.generateSeed(b);
+                engineSetSeed(b);
+            }
+        } finally {
+            objLock.unlock();
+        }
+    }
+
+    @Override
+    public int getPriority() {
+        if (isSeedGenerator)
+            return -12;
+        return -11;
```

Comment:
How are priorities determined?  I see other implementors of this method mostly setting their priority to 0.

Reply:
Hi @DanHeidinga 
Default global CRaC Context implementation uses resource priorities to define the order of checkpoint notifications (beforeCheckpoint). Resource priorities are automatically assigned incrementally upon resource registration in the context. The lowest priority is 0. Order of checkpoint notification is the reverse order of registration.
There are JDKContext and JDKResource internal implementations. These classes allow assigning priorities to specific resources. The order of checkpoint notifications for JDKResources is from higher to lower priorities.

Reply:
Given that explanation, how were `-12` & `-11` chosen here?  Randomly as low enough to ensure these are changed late?

Reply:
Yes and No :) I left some space for the future JDK resources that depend on the secure randoms (use secure random inside own beforeCheckpoint method)
Maybe these priorities should be collected in a single class, e.g. defined in the JDKresource interface

Reply:
Added pull request to collect JDKResorce priorities in the single enumeration: https://github.com/openjdk/crac/pull/8
Current PR will be updated as soon as JDKResource priorities is accepted and comitted